C++: Refactor SSA usage in data flow.

aschackmull opened the pull request:
Commit-by-commit review encouraged.

There are a bunch of minor commits, and then there's "C++: Use SSA data flow integration module", which is the main change that switches C++ from direct phi-read references to use the data flow integration module to construct use-use flow.

I've done a lot of local black-box testing of these changes and they generally looking very reasonable. There are minor changes on the order of 0.1% of the ssa-based data flow edges, but at this point I think further evaluation is best done via dca.

MathiasVP commented on shared/ssa/codeql/ssa/Ssa.qll:
```diff
@@ -1599,6 +1599,13 @@ module Make<LocationSig Location, InputSig<Location> Input> {
+       *
+       * Gets the basic block index of this node.
+       */
+      abstract int getIndex();
```
This also needs to be implemented for the other languages, right? Otherwise, I assume compilation will fail?

MathiasVP reviewed (commented):
This looks great! If tests DCA comes back happy I'm perfectly happy merging this and leaving the small TODOs.

Update: I see CI is already finding some test discrepancies. Happy to take a look at those

MathiasVP commented:
The result changes should be fixed by https://github.com/github/codeql/pull/18955

aschackmull commented:
Rebased and added two new commits. The first commit "C++: Fix spurious ExprNode fanout in DataFlowIntegration" should improve performance by using proper control-flow nodes instead of `UseImpl` as the instantiation of `Expr` in `DataFlowIntegration`. A side-effect of that is that barrier-guards can no longer properly identify guarded uses, so that part of barrier-guards is moved to the C++ side.
The second commit "C++: Remove superfluous disjunct" is a very minor cleanup. The second disjunct is completely contained in the first and thus superfluous.

MathiasVP commented:
And finally: The lost test result in the internal directory will be fixed by https://github.com/github/codeql/pull/19001

MathiasVP commented:
## Analysis of result changes

### neovim__neovim
- `cpp/invalid-pointer-deref`: The two lost results are TPs. It's caused by the change to the barriers in the query. I still need to fully understand which barrier is now being identified that wasn't before, but I'm not too comfortable with this change yet.

MathiasVP commented:
Overall, the changes in code and results LGTM. It's a bit of a shame that we lose those two `cpp/invalid-pointer-deref`, and I don't know how to regain them without exposing phi (reads) 😭. But I think it's a price I'm willing to pay for more code sharing.

Since this is a fairly major change (and I'm not actually on the team that owns this code anymore!) I would appreciate it if someone from the C team would take a second look. I'll ping them on Slack
